python/ray/data/_internal/logical/rules/zero_copy_map_fusion.py

```diff
@@ -0,0 +1,79 @@
+from abc import abstractmethod
+from typing import List
+
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.map_transformer import (
+    BlockMapTransformFn,
+    BuildOutputBlocksMapTransformFn,
+    MapTransformFn,
+    MapTransformFnDataType,
+)
+from ray.data._internal.logical.interfaces.optimizer import Rule
+from ray.data._internal.logical.interfaces.physical_plan import PhysicalPlan
+
+
+class ZeroCopyMapFusionRule(Rule):
+    """Base class for zero-copy map fusion rules.
+
+    Subclasses implement the optimization strategies for different combinations of
+    fused map operators, by dropping unnecessary data conversion `MapTransformFn`s.
+    """
+
+    def apply(self, plan: PhysicalPlan) -> PhysicalPlan:
+        self._traverse(plan.dag)
+        return plan
+
+    def _traverse(self, op):
+        if isinstance(op, MapOperator):
+            map_transformer = op.get_map_transformer()
+            transform_fns = map_transformer._transform_fns
+            new_transform_fns = self._optimize(transform_fns)
+            # Physical operators won't be shared,
+            # so it's safe to modify the transform_fns in place.
+            map_transformer._transform_fns = new_transform_fns
+
+        for input_op in op.input_dependencies:
+            self._traverse(input_op)
+
+    @abstractmethod
+    def _optimize(self, transform_fns: List[MapTransformFn]) -> List[MapTransformFn]:
+        """Optimize the transform_fns chain of a MapOperator.
+
+        Args:
+            transform_fns: The old transform_fns chain.
+        Returns:
+            The optimized transform_fns chain.
+        """
+        pass
+
+
+class ReadOpZeroCopyMapFusion(ZeroCopyMapFusionRule):
```

Comment:
Can we be more general for the naming? Such as `EliminateBuildOutputblocks`? There's no read/write operator in physical world, and this optimization rule is really doing the work to eliminate unnecessary `BuildOutputBlocksMapTransformFn`.

Reply:
+1